### ISSUE TITLE
Getting parameterised query text without placeholders

### ISSUE THREAD

**Issue by GoldAndPurple** [jtv/libpqxx#902]
Is there a way to know the proper text of the executed query, with placeholder text replaced with bound parameters? Relevant for exception handling and logging purposes, where you might not know the parameters used, but need information the most. 
Right now there is a weird pattern for more complicated queries. It's better to format query-strings with escaped parameters in the application itself, "on your own", and only then execute the resulting statement with no parameters.

**Comment by jtv** [comment]
Alas, I don't know if any way to get that version of the query string.

I guess it would be possible to reconstruct it, but that can go wrong in exactly the kind of situation where you need it most.

**Comment by GoldAndPurple** [comment]
I suppose the main problem is that postgres itself doesn't always log parameters, so we have to rely on either the library, or the application to remember them.
But what are the pitfalls of reconstructing the query post-execution? The only thing that wouldn't be exactly the same is raw binary data (and its user-defined equivalents), right? Everything else is passed as text, so what's the problem with replacing placeholders with pqxx-escaped parameters?

**Comment by jtv** [comment]
Text is no longer simple.  :-(  Yes, it can be _done,_ but it's starting to sound like _work._

Imagine you're working in an encoding like SJIS.  Your parameters look like `$1`, `$2`, etc.  But your query can also contain a multi-byte character where the 2nd or 3rd byte happens to match the ASCII byte value for `$`.  Maybe it' s a Japanese multi-byte character that libpqxx doesn't support properly yet, or there's an obscure bug in text parsing.  Now you may have a phantom parameter, and the parameter value may splice into the middle of a singe character.

Or consider the complexity of support for complex data types — support that may live in your application rather than in libpqxx itself.  I'm not sure the quoting and escaping rules will always be appropriate.  Will libpqxx need to figure out the parameters' SQL types?  How do we keep those choices consistent with libpq's?  And how do we represent binary?

In theory it's all doable.  But it does add complexity and scope for error, and error handling is the worst place for that.

What might work better is to show the query as-is, and then also make the parameter values available.  However I do kind of hate to leave that responsibility on your shoulders because it exposes your application to encoding issues.  For instance, you might be running on Windows in UTF-16, but with UTF-8 as your postgres client encding — that probably means you can't just print these values.  And you may want to truncate long parameter values for display, but you'll need to take care not to truncate in the middle of a character.  (And then in some languages, there are places where truncating between characters doesn't make sense in some cases, but that's a minor thing so let's not even get into that.)

In short: I hate parsing text in dynamically determined encodings.  :-)

**Comment by jtv** [comment]
Looking at this again now, it all seems a bit less dramatic.  What we can probably do is leave the query text as-is, but _append_ some text describing the parameters.  In fact I think that would be even more helpful.

The other thing though is that I don't know when I might get around to this.  I'm finally doing serious work on libpqxx 8.0, which will have much more far-reaching consequences for exceptions.  So it may be quite some time.  I'll keep the ticket open.

**Comment by jtv** [comment]
Sometimes I wish I could tag an issue like this as _back-burner_ to tell it that I'll get around to it eventually, and it doesn't have to be so aggressive in looking for inactivity.

**Comment by tt4g** [comment]
The `exempt-issue-labels` option may be used.

https://github.com/actions/stale/blob/v9.1.0/README.md#exempt-issue-labels

> #### exempt-issue-labels
> 
> Comma separated list of labels that can be assigned to issues to exclude them from being marked as stale
> (e.g: `question,bug`)
> 
> If unset (or an empty string), this option will not alter the stale workflow.
> 
> Default value: unset

Several other `exempt-*` options are provided.

**Comment by jtv** [comment]
Ah @tt4g you've mentioned this before and I never got around to saying — how or where do I use that label?  This README looks to me like it documents some kind of tool or plugin that has that option.

**Comment by tt4g** [comment]
@jtv Add `exempt-issue-labels` to the `with` dictionary in `actions/stale@v1`.
Issues with the labels listed here will no longer be marked as stale.

https://github.com/jtv/libpqxx/blob/e67535418db37b36155b2ea39a5e3c18a4b6001c/.github/workflows/stale.yml#L19-L30

An example of excluding issues with the “enhancement” or “bug” labels:

```diff,yaml
     steps:
       - uses: actions/stale@v1
         with:
           repo-token: ${{ secrets.GITHUB_TOKEN }}
           stale-issue-message: \
             There has been no activity on this ticket.  Consider closing it.
           stale-pr-message: \
             There has been no activity on this pull request. \
             Complete it or drop it.


           stale-issue-label: 'no-issue-activity'
           stale-pr-label: 'no-pr-activity'
+          exempt-issue-labels: 'enhancement,bug'
```

By the way, after seeing your comment on “There has been no activity on this ticket. Consider closing it,” I assumed you wanted to suppress GitHub Action `actions/stale`, which is why I made that comment.

**Comment by jtv** [comment]
Thanks @tt4g  — I've created a dedicated label for this.
